File: Java/hash_table/HashTable.java
package hash_table;

public class HashTable{
    private int[] table;
    private int mode;
    private int numCollisions;
    private int numElements;
    private int tableSize;
    private double aValue;

    // Constructors
    public HashTable(int mode){
        this.mode = mode;
        this.numCollisions = 0;
        this.numElements = 0;
        this.tableSize = 1;
        this.table = new int[this.tableSize];

        if(mode == 3){
            this.aValue = Math.random();
        }
    }


    public HashTable(int mode, int tableSize){
        this.mode = mode;
        this.numCollisions = 0;
        this.numElements = 0;
        this.tableSize = tableSize;
        this.table = new int[this.tableSize];

        if(mode == 3){
            this.aValue = Math.random();
        }
    }


    // Private Methods
    private int numDigits(){
        int tempSize = this.tableSize - 1;
        int r = 0;
        while (tempSize > 0){
            tempSize /= 10;
            r++;
        }

        return r;
    }


    private int divHash(int key){
        int idx = key % this.tableSize;
        if(this.table[idx] != 0){
            this.numCollisions++;
            idx = this.divHash(key + 1);
        }
        
        return idx;
    }


    private int midSquareHash(int key){
        int idx = 0;
        int r = this.numDigits();

        String square = String.valueOf(key * key);
        idx = Integer.parseInt(square.substring((square.length() / 2) - (r / 2), r));

        if(this.table[idx] != 0){
            this.numCollisions++;
            idx = this.midSquareHash(key + 1);
        }

        return idx;
    }


    private int digitFoldingHash(int key){
        int idx = 0;
        int r = this.numDigits();

        String keyStr = String.valueOf(key);
        while(keyStr.length() > 0){
            if(r > keyStr.length()){
                r = keyStr.length();
            }

            idx += Integer.parseInt(keyStr.substring(0, r));
            keyStr = keyStr.substring(r);
        }

        if(this.table[idx] != 0){
            this.numCollisions++;
            idx = this.digitFoldingHash(key + 1);
        }

        return idx;
    }


    private int multHash(int key, double a){
        int idx = (int)(this.tableSize * ((key * a) % 1));

        if(this.table[idx] != 0){
            this.numCollisions++;
            idx = this.multHash(key + 1, a);
        }

        return idx;
    }


    private void resize(){
        HashTable newTable = new HashTable(this.mode, this.tableSize * 2);
        for(int i = 0; i < this.tableSize; i++){
            if(this.table[i] != 0){
                newTable.insert(this.table[i]);
            }
        }

        this.table = newTable.table;
        this.tableSize = newTable.tableSize;
        this.numCollisions = newTable.numCollisions;
        this.numElements = newTable.numElements;
        this.aValue = newTable.aValue;
    }


    // Public Methods
    public void insert(int key){
        if((double)this.numElements / (double)this.tableSize > 0.7){
            this.resize();
        }

        int idx;
        switch(this.mode){
            case 0:
                idx = this.divHash(key);
                break;
            case 1:
                idx = this.midSquareHash(key);
                break;
            case 2:
                idx = this.digitFoldingHash(key);
                break;
            case 3:
                idx = this.multHash(key, this.aValue);
                break;
            default:
                idx = this.divHash(key);
                break;
        }

        this.table[idx] = key;
        this.numElements++;
    }
}

File: JavaScript/hash_table/hash_table.js
class HashTable {
    // Constructors
    constructor(mode) {
        this._mode = mode;
        this._numCollisions = 0;
        this._numElements = 0;
        this._tableSize = 1;
        this._table = new int[this.tableSize];

        if (mode == 3) {
            // Assign a_value to a random double between 0 and 1
        }
    }

    constructor(mode, tableSize) {
        this._mode = mode;
        this._numCollisions = 0;
        this._numElements = 0;
        this._tableSize = tableSize;
        this._table = new int[this.tableSize];

        if (mode == 3) {
            // Assign a_value to a random double between 0 and 1
        }
    }

    // Private Methods
    _numDigits() {
        let tempSize = this._tableSize - 1;
        let r = 0;
        while (tempSize > 0) {
            tempSize /= 10;
            r++;
        }

        return r;
    }


    _divHash(key) {
        let idx = key % this._tableSize;
        if (this._table[idx] != 0) {
            this._numCollisions++;
            idx = this._divHash(key + 1);
        }

        return idx;
    }


    _midSquareHash(key) {
        let r = this._numDigits();

        const square = (key * key).toString();
        let idx = Number(square.substring((square.length / 2) - (r / 2), r));

        if (this._table[idx] != 0) {
            this._numCollisions++;
            idx = this._midSquareHash(key + 1);
        }

        return idx;
    }


    _digitFoldingHash(key) {
        let idx = 0;
        let r = this.num_digits();

        let keyStr = key.toString();
        while (keyStr.length > 0) {
            if (r > keyStr.length) {
                r = keyStr.length;
            }

            idx += Number(keyStr.substring(0, r));
            keyStr = keyStr.substring(r);
        }

        if (this._table[idx] != 0) {
            this._numCollisions++;
            idx = this._digitFoldingHash(key + 1);
        }

        return idx;
    }


    _multHash(key, a) {
        let idx = (int)(this._tableSize * ((key * a) % 1));

        if (this._table[idx] != 0) {
            this._numCollisions++;
            idx = this._multHash(key + 1, a);
        }

        return idx;
    }


    _resize() {
        const new_table = new HashTable(this._mode, this._tableSize * 2);
        for (let i = 0; i < this._tableSize; i++) {
            if (this._table[i] != 0) {
                new_table.insert(this._table[i]);
            }
        }

        this = new_table;
    }

    // Public Methods
    insert(key) {
        if (this._numElements / this._tableSize > 0.7) {
            this._resize();
        }

        let idx = 0;
        switch (this._mode) {
            case 0:
                idx = this._divHash(key);
                break;
            case 1:
                idx = this._midSquareHash(key);
                break;
            case 2:
                idx = this._digitFoldingHash(key);
                break;
            case 3:
                idx = this._multHash(key, this._a);
                break;
            default:
                idx = this._divHash(key);
                break;
        }

        this._table[idx] = key;
        this._num_elements++;
    }
}